### PR TITLE
win_psrepository - add script and publish options, resolve locations

### LINKED ISSUE #67501: Add script and publish location support to win_psrepository
##### SUMMARY

The `win_psrepository` module only supports the `source` (module source) location, which means that the repositories it registers can't be used for downloading scripts, or publishing either modules or scripts.

Ideally it would add the remaining 3 location types so that complete repositories can be used.

The other 3 (PowerShell names) are:
- `PublishLocation`
- `ScriptSourceLocation`
- `ScriptPublishLocation`

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

lib/ansible/modules/windows/win_psrepository.ps1

##### ADDITIONAL INFORMATION

Example usage:

```yaml
win_psrepository:
  name: MyRepo
  script_source_location: https://myrepo.local/scripts
  publish_location: https://myrepo.local/module_publish
  script_publish_location: https://myrepo.local/script_publish
```

### LINKED ISSUE #67504: win_psrepository - perpetually "changed" if source location redirects
##### SUMMARY

If you register a repository where the URL redirects to a new URL, then PowerShell will follow the redirects and register the final URL in the repository. 

This can also happen because PowerShellGet attempts to figure out if NuGet repositories support api v2 or v3 and it may add additional api version paths to the URL.

The ansible module doesn't know this so on the next run, it will compare the existing repository's (now modified) source URL to the one supplied in options, see that it's different, and "reset" it, but the end result will be the same, so the status will always be `changed` unless the playbook is updated.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

lib/ansible/modules/windows/win_psrepository.ps1

##### ANSIBLE VERSION

```paste below
ansible 2.9.4
  config file = None
  configured module search path = ['/home/briantist/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.5/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.5.2 (default, Oct  8 2019, 13:06:37) [GCC 5.4.0 20160609]
```

##### CONFIGURATION

```paste below
```
(no config changes)

##### OS / ENVIRONMENT
 any Windows

##### STEPS TO REPRODUCE

Register any repository that includes a redirect. It doesn't have to be a real repository to reproduce the problem, just any URL that redirects.

```yaml
win_prespository:
  name: MyRepo
  source: https://ansible.com
```
(`https://ansible.com` will redirect to `https://www.ansible.com`)

Run once, run again.

Also can check repository manually:
```powershell
Get-PSRepository MyRepo
```

##### EXPECTED RESULTS

Status `OK` on second run.

##### ACTUAL RESULTS

Status is `CHANGED`.

### PR DESCRIPTION
**Pull request by briantist**
##### SUMMARY

* Add parameters for script and publish locations ( Fixes #67501 )

* Allow for forced update
  * Normal mode of operation on existing repositories is to update only the settings you supplied (same as previously)
  * `force` will make the repository match your settings exactly, by unregistering and re-registering the repository if there are changes
  * some desired changes can't be made without `force`, like if you wanted to "unset" a publish location after having previously set it to any value; there's no way to do that without registering anew.
* Add new tests for above.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

 lib/ansible/modules/windows/win_psrepository.ps1

##### RELATED PRs:
- #67674 
- #67594

### CODE REVIEW

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_psrepository.ps1]: That's a lot of aliases, I would try to just keep `source` as backwards compatibility and make this the `source_location` option. Because the underlying cmdlet is also `SourceLocation` I don't think having the `module_*` ones should matter too much as long as we document the details of this option in the description.

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_psrepository.ps1]: Same thing here, I would probably just have `publish_location` as that is the default type and only have the `script_*` ones with the prefix.

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_psrepository.ps1]: [Resolve-Location](https://github.com/PowerShell/PowerShellGet/blob/development/src/PowerShellGet/private/functions/Resolve-Location.ps1) looks like it might be a private function I would be very reluctant to rely on that in our modules. What is it meant to be doing? Trying to get the final URL on a redirected one?

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_psrepository.ps1]: You have a lot of common code here, might be best to put in a loop or in your own cmdlet.

File: test/integration/targets/win_psrepository/defaults/main.yml
Comment (jborean93) [inline test/integration/targets/win_psrepository/defaults/main.yml]: httpbin.org is notoriously temperamental in CI and causes all sorts of unreachable errors when we run it. To combat this issue we have a test container that runs the httpbin code locally. What you need to do to access this is

* Add the `needs/httptester` in your test target alias like you can see with the [win_get_url](https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_get_url/aliases) aliases (you don't need to skip windows 2008 as we no longer test that version)
* Add `prepare_http_tests` in the test target meta/main.yml. This sets things up like certificates and gets it ready for cleanup once the test is finished
* In your tests/variables change any reference `http://httbin.org/<path>` to `http://{{ httpbin_host }}/<path>`. This is automatically set to the private instance if available or just `httpbin.org` if httptester is not available

The beauty of httptester is that it is reliable, a private instance, has trusted https certs, and if ansible-test can't set it up the `{{ httpbin_host }}` var just points to the public instance for your own testing.

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_psrepository.ps1]: These statements don't do anything when run in PowerShell. If you wanted to provide a nice error in case the requirements isn't met then you need to do a manual check.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (briantist) [reply lib/ansible/modules/windows/win_psrepository.ps1]: It was to allow for consistency in naming for those who want it, but point taken, I don't feel strongly about keeping them.

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (briantist) [reply lib/ansible/modules/windows/win_psrepository.ps1]: Yes it's private, I wrote info about it in the XML help.

I originally built a function that followed HTTP/S redirects. It was working for that purpose but I found that it failed to produce the same URL as PowerShellGet in some cases and that's when I found out that this internal function tries to do other stuff, like probe for NuGet API responses and add on versioned API extensions to the URL. It also does some minor changes to paths in filesystem repos. All in all, it seemed like an awful lot to be duplicating the logic of it all. 

I think relying on the internal function gives us the most consistent outcome, especially if that function changes in future versions as the result of `Resolve-Location` will always be what PowerShellGet itself is writing out. 

The next best alternative is probably to copy and paste it out of the source code and duplicate it here, which protects us from the name changing, but leaves us more vulnerable to the implementation changing.

Open to suggestions on alternatives

File: test/integration/targets/win_psrepository/defaults/main.yml
Comment (briantist) [reply test/integration/targets/win_psrepository/defaults/main.yml]: Ah, that's fantastic! Really appreciate the detailed instructions there.

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (jborean93) [reply lib/ansible/modules/windows/win_psrepository.ps1]: There's nothing stopping PSGet from changing/moving/removing this cmdlet in future updates effectively breaking this module. If you wanted to fix this particular problem I would just try and replicate the same behaviour as PSGet and doing a GET (HEAD might even be better) and returning the location if it was redirected https://github.com/PowerShell/PowerShellGet/blob/development/src/PowerShellGet/private/functions/Ping-Endpoint.ps1#L107.

You can even utilise the [Ansible.ModuleUtils.WebRequest](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/powershell/Ansible.ModuleUtils.WebRequest.psm1). You don't need to expose the whole share parameters aspect but you can utilise `Get-AnsibleWebRequest` and `Invoke-AnsibleWebRequest` with `-IgnoreBaseResponse` to do a similar thing as PSGet. Feel free to reach out to me if you need some more help on trying to get that set up. We do something similar in win_uri so that's a good place to start. If you do go down that route I wouldn't expose those parameters as module options, just pass in whatever you need normally.

Comment (briantist) [reply lib/ansible/modules/windows/win_psrepository.ps1]: I really did go down that road before ([commit where it's removed is here](https://github.com/ansible/ansible/pull/67499/commits/0b3d4c6aae02892e6169ceea11a68cdb01f0a376)), even used `HEAD` requests, until it broke in the tests because myget.org doesn't support `HEAD` requests, so I added fallback to `GET`, and continued on until running into the next issue which is that `Resolve-Location` doesn't just follow 3xx redirects. So there's more to add if we were going to try to emulate that behavior. The function I had worked fine for purely HTTP/S redirects.

Highly recommended reading as this shed a lot of light on what a minefield this is: https://github.com/PowerShell/PowerShellGet/issues/317

If it's more desirable to re-implement their logic than copying their code directly (it's MIT licensed) or relying on the private method, I think it might be a better idea to leave that to a new PR and keep that as a separate issue. 

My main purpose was to add support for the other location parameters; I just happened to find this bug during the course of that, but it's turned out to be far more time consuming than expected to address and is probably best solved on its own.

What do you think of a compromise for the time being where if the private method is not found or fails, we issue a warning and continue on? The result in that scenario would be the same behavior that's in place now (but with notice) in event that something internal changes, but in the meantime it would be more reliable (this would still be leaving #67504 open so that it can be addressed permanently in a new PR).

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (briantist) [reply lib/ansible/modules/windows/win_psrepository.ps1]: Ok I'll see what I can do to clean that up

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (jborean93) [reply lib/ansible/modules/windows/win_psrepository.ps1]: Totally understand that this is not an ideal situation all round and PowerShellGet has a lot of quirks that make these modules difficult to deal with. In saying that by trying to re-utilise the `Resolve-Location` cmdlet which isn't even exposed in the modules manifest you are relying on an internal implementation that is not guaranteed to be there in future releases. Considering there is a v3 release in the works that is meant to solve a lot of reliability issues of PSGet I would wager that this function has gone the way of the dodo effectively breaking our module if we took the changes as is.

I wouldn't want to copy their code, I've been burnt too many times with PSGet to consider having their code in these modules and even just trying to decipher their current implementation is a bit mind numbing. Ultimately I think if you wanted to try and solve this particular issue (and if I understand the problem correctly) you can solve it with something like

```
Function Resolve-Location {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory=$true)]
        [Uri]
        $Location,

        [Parameter(Mandatory=$true)]
        [Object]
        $Module
    )

    if ($Location.Scheme -notin @('http', 'https')) {
        $Location
        return
    }

    $request = Get-AnsibleWebRequest -Uri $Location -Method GET -FollowRedirects All
    try {
        Invoke-WithWebRequest -Request $request -IgnoreBadResponse -Module $Module -Script {
            Param ([System.Net.WebResponse]$Response, [System.IO.Stream]$Stream)
    
            $Response.ResponseUri.AbsoluteUri
        }
    } catch [System.Net.WebException] {
        # Probably unable to resolve the URL, just ignore this and rely on PSGet to fail or not.
        $Location
    }
}

$source_location = Resolve-Location -Location $source_location -Module $module
$publish_location = Resolve-Location -Location $publish_location -Module $module
...
```

By using this util you take advantage of all the complexities when it comes to dealing web requests in PowerShell such as

* Remembering to enable TLS 1.2
* Easily set the timeout or redirection settings if desired
* If we wanted to add support for proxies or web auth, this is all tested and done for you in the util

A quick test shows that this work for me and seems to be similar to what PSGet does, just with less lines.

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (jborean93) [reply lib/ansible/modules/windows/win_psrepository.ps1]: Oh and just after looking at, these options should try and match up with the return values there. When looking at the result for just 1 repo I get

```
ok: [2019] => {
    "changed": false,
    "invocation": {
        "module_args": {
            "name": "*"
        }
    },
    "repositories": [
        {
            "installation_policy": "Untrusted",
            "name": "PSGallery",
            "package_management_provider": "NuGet",
            "provider_options": {},
            "publish_location": "https://www.powershellgallery.com/api/v2/package/",
            "registered": true,
            "script_publish_location": "https://www.powershellgallery.com/api/v2/package/",
            "script_source_location": "https://www.powershellgallery.com/api/v2/items/psscript",
            "source_location": "https://www.powershellgallery.com/api/v2",
            "trusted": false
        }
    ]
}
```

So having `source_location` and `publish_location` as is but keep the prefix for the script ones.

File: lib/ansible/modules/windows/win_psrepository.ps1
Comment (briantist) [reply lib/ansible/modules/windows/win_psrepository.ps1]: That solves for 3xx redirects but not the other modifications that `Resolve-Location` does (v2/v3 API checks, filesystem path changes). 

I definitely agree that relying on a private function is not ideal and may be prone to failure, so I understand not wanting to take a dependency on it.

Given the complexity I think I'd like to separate this issue into a new PR that directly addresses it (maybe after the split), and keep this PR scoped down to adding the new locations and `force` support.

In the new PR I can work toward using the Ansible web request module util like you have above.

Comment (jborean93) [reply lib/ansible/modules/windows/win_psrepository.ps1]: That sounds like a great plan. Let’s get in support for the new parameters and not have this hold up the PR.

**Comment by briantist** [comment]
ready_for_review

**Comment by jborean93** [comment]
PR has been migrated to https://github.com/ansible-collections/community.windows/pull/11.

**Pull request closed without merging**
